src/Routing/Route/Route.php

```diff
@@ -131,6 +131,20 @@ class Route
      */
     public function __construct(string $template, array $defaults = [], array $options = [])
     {
+        $checker = function () use ($defaults): void {
+            foreach (['plugin', 'prefix', 'controller', 'action'] as $key) {
+                if (isset($defaults[$key]) && !is_string($defaults[$key])) {
+                    trigger_error(
+                        'Value for `' . $key . '` in $defaults when connecting routes'
+                        . ' must be of type `string` or `null`',
+                        E_USER_ERROR
+                    );
```

Comment:
Why not raise an exception instead of triggering errors?

Reply:
Especially since you closed https://github.com/cakephp/cakephp/pull/17927 which would have prevented a hard TypeError fail here
So any non matching type would blow up hard along the line now anyway.